[EA] Heart banner for Donation Celebration

Will-Howard opened the pull request.

Will-Howard commented on packages/lesswrong/server/repos/ForumEventsRepo.ts:
```diff
@@ -34,6 +34,25 @@ class ForumEventsRepo extends AbstractRepo<"ForumEvents"> {
     `, [userId, _id])
   }
+
+  // TODO Add format + version to the data, so it can be distinguished from votes
```
This is intentionally left here

Will-Howard commented on packages/lesswrong/components/forumEvents/GivingSeason2024Banner.tsx:
```diff
@@ -20,6 +20,14 @@ import type { ForumIconName } from "../common/ForumIcon";
 
 const DOT_SIZE = 12;
 
+/** Style to pass pointer events through one element, but not have this affect all children */
+const PASS_THROUGH_POINTER_EVENTS = {
+  pointerEvents: "none",
+  "& > *": {
+    pointerEvents: "all",
+  },
+}
+
```
One of the trickier parts of this was making it so that you can't place a heart on elements covering part of the banner (so you can still use these elements). This required setting all the z indexes the same and carefully managing which elements capture pointer events

Will-Howard replied:
Ah, well seeing as this is working now I'm going to leave it as is, but this is a good idea for if managing this pass-through thing becomes too complicated (e.g. when we want to use this for a non-giving-season banner)

agnesstenlund commented:
This looks so good!! Thanks @Will-Howard 🙌 

I only have a few nitpicks:

- [ ] In dark mode I think you can't see the border around the input when adding the note?
- [ ] Not sure if the text below "Donation Celebration Week" will automatically update on the banner when the week starts. We should probably change to "Click to add a heart to the banner showing that you’ve done your annual donation" when it starts
- [ ] I'm assuming the "+ Place heart" is there because you can't just click anywhere on mobile? I think we may want to change it to "Add heart" (since that's the language we use elsewhere, and matches the +), and maybe we also want to make it look more like a button? Not quite sure yet. Will have a try

Also two questions:
1. Why have the ability to pin comments? What does that do?
2. During debate week, did we show your own comment similar to this? Or with your avatar showing?

oetherington reviewed (approved):
Seems good - I left a few comments.

There's also a slightly annoying bug where the first time you add a heart after a page load all the hearts dissappear for a second before coming back. In general, I think it'd also be good to keep a copy of the users heart coordinates in a `useState` on the client because the network lag after adding a heart is pretty jarring (and it'd probably also solve the other bug I mentioned).

Will-Howard commented:
@agnesstenlund 

> Not sure if the text below "Donation Celebration Week" will automatically update on the banner when the week starts. We should probably change to "Click to add a heart to the banner showing that you’ve done your annual donation" when it starts

Fixed 👍 .

> I'm assuming the "+ Place heart" is there because you can't just click anywhere on mobile? I think we may want to change it to "Add heart" (since that's the language we use elsewhere, and matches the +), and maybe we also want to make it look more like a button? Not quite sure yet. Will have a try

👍 I've changed it to "+ Add heart". For the message that said "...placing heart (click to cancel)", this is actually not anything to do with something loading on the site, it's just to give the user an ability to cancel placing the heart if they click accidentally. I've changed it to this to hopefully make it clearer:

<img width="501" alt="Screenshot 2024-12-19 at 14 36 44" src="https://github.com/user-attachments/assets/abd8380f-017f-4f3f-adb3-7847b083cae3" />

> The x is slightly misaligned

Should be fixed:

<img width="81" alt="Screenshot 2024-12-19 at 14 38 45" src="https://github.com/user-attachments/assets/8b5cc7c0-f123-4bd5-9e07-f7f51231a81e" />

> we may want to change "Cancel" to "Skip for now" instead

Done, I changed it to just "skip" to avoid taking up too much space

<img width="400" alt="Screenshot 2024-12-19 at 14 46 41" src="https://github.com/user-attachments/assets/cbeb9214-1bc3-4a4f-9f13-3ff91de1b4d8" />

> I couldn't edit my comment, I think?

There is no way to edit it from that tooltip view, but you can do so from the thread. I think the easiest change here would be to change some copy to nudge people towards going to the thread to do anything complicated. Any suggestions?

> This could be a problem of the inspector, but I have to wait for a long time here:
...
> I find that pattern a bit confusing 👆 , especially since clicking otherwise is the thing you do to add it (I want to click to add at this point). Did we add it since it takes a while to load? Otherwise I'd just show a spinning thing and add it? (Without the "Click to cancel")

These two should be resolved by this (from above): "For the message that said "...placing heart (click to cancel)", this is actually not anything to do with something loading on the site, it's just to give the user an ability to cancel placing the heart if they click accidentally. I've changed it to this to hopefully make it clearer: (screenshot)". Do you still think it's confusing?

> Why have the ability to pin comments? What does that do?

It makes it so you can move your mouse away without the comment disappearing. I think the case where this is most useful is if you are replying (in which case it pins itself automatically). This was borrowed from the debate week poll version, and I think maybe Toby requested it there? I did find it useful in that case for reading long comments

> During debate week, did we show your own comment similar to this? Or with your avatar showing?

This is what it looked like then (sorry fully expanded was the only image I could find easily):
![image](https://github.com/user-attachments/assets/4511872e-07b5-4f72-9999-2b32320a130b)

Your avatar appeared separately on the bottom slider as well as in the results, so it was easier to find your own comment in that case. You could also edit your comment from the slider, it showed a different view with more info than the hover version of the comments

Will-Howard commented on packages/lesswrong/components/forumEvents/GivingSeason2024Banner.tsx:
```diff
@@ -548,7 +588,9 @@ const GivingSeason2024Banner = ({classes}: {
             className={classNames(classes.background, name === selectedEvent.name && classes.backgroundActive)}
-          />
+          >
+            {name === "Donation Celebration" && <ForumEventStickers interactive={name === selectedEvent.name} />}
```
> I think we only want to render this when the current event is "Donation Celebration", right?

Yep that's right, good catch!

Will-Howard commented on packages/lesswrong/components/forumEvents/ForumEventResultPopper.tsx:
```diff
@@ -0,0 +1,228 @@
+                    icon="Pin"
+                    onClick={() => setIsPinned(!isPinned)}
+                    className={classNames(classes.pinIcon, { [classes.pinIconPinned]: isPinned })}
+                  />
```
It makes it so you can move your mouse off the tooltip and have it stay open. It appears to be working for me

agnesstenlund commented:
Sorry totally missed the notification about it being done 🤦‍♀️ but looks good, happy you merged it!
